### PR TITLE
feat(captcha): captcha guard and module

### PR DESCRIPTION
**Pull request by ABCxFF**

### CODE REVIEW

**Review by NathanFlurry — CHANGES_REQUESTED** [review]
the logic is incorrect for captchas. captchas are supposed to require a captcha on the first request.
this module needs better doc comments about all of the scripts.
the scripts are missing description.
the module is missing an icon & tags & description.
don't commit the deno.lock. this might be an issue with the toolchain.
make sure to run rivet module format and rivet module lint

**Review by NathanFlurry — CHANGES_REQUESTED** [review]
- the logic is incorrect for captchas. captchas are supposed to require a captcha on the first request.
- this module needs better doc comments about all of the scripts.
- the scripts are missing description.
- the module is missing an icon & tags & description.
- don't commit the deno.lock. this might be an issue with the toolchain.
- make sure to run rivet module format and rivet module lint

File: modules/captcha/utils/providers/hcaptcha.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+const API = "https://api.hcaptcha.com/siteverify";
+export const validateHCaptchaResponse = async (
+    secret: string,
+    response: string
+): Promise<boolean> => {
+    try {
+        const body = new FormData();
+        body.append("secret", secret);
+        body.append("response", response);
+        const result = await fetch(API, {
```
Comment (NathanFlurry) [inline modules/captcha/utils/providers/hcaptcha.ts:10]: this doesn't validate response status code

File: modules/captcha/utils/providers/turnstile.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+const API = "https://challenges.cloudflare.com/turnstile/v0/siteverify";
+export const validateCFTurnstileResponse = async (
+    secret: string,
+    response: string
+): Promise<boolean> => {
+    try {
+        const result = await fetch(API, {
```
Comment (NathanFlurry) [inline modules/captcha/utils/providers/turnstile.ts:7]: this doesn't validate response status code

File: modules/captcha/utils/providers/turnstile.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+            }
+        });
+
+        const { success } = await result.json();
+
+        return success;
+    } catch {}
```
Comment (NathanFlurry) [inline modules/captcha/utils/providers/turnstile.ts:18]: all errors should be logged

File: modules/captcha/utils/providers/hcaptcha.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+            body,
+            method: "POST",
+        });
+
+        const { success } = await result.json();
+
+        return success;
+    } catch {}
```
Comment (NathanFlurry) [inline modules/captcha/utils/providers/hcaptcha.ts:18]: all errors should be logged

File: modules/captcha/scripts/verify_captcha_token.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import { RuntimeError, ScriptContext } from "../module.gen.ts";
+import { validateHCaptchaResponse } from "../utils/providers/hcaptcha.ts";
+import { validateCFTurnstileResponse } from "../utils/providers/turnstile.ts";
+// import { validateHCaptchaResponse } from "../providers/hcaptcha.ts";
```
Comment (NathanFlurry) [inline modules/captcha/scripts/verify_captcha_token.ts:4]: remove unused comments

**Review by NathanFlurry — CHANGES_REQUESTED** [review]

**Comment by ABCxFF** [comment]
* **#581** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/581?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#578** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/578?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#577** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/577?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#576** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/576?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ABCxFF and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ABCxFF** [comment]
* **#581** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/581?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#578** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/578?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#577** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/577?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#576** <a href="https://app.graphite.dev/github/pr/rivet-gg/modules/576?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ABCxFF and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: modules/captcha/scripts/verify_captcha_token.ts (hunk 5 above)
Comment (ABCxFF) [reply modules/captcha/scripts/verify_captcha_token.ts:4]: Resolved at top of stack 5/5

**Pull request closed without merging**